src/Microsoft.AspNetCore.Diagnostics.HealthChecks/HealthCheckMiddleware.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.Diagnostics.HealthChecks
+{
+    public class HealthCheckMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HealthCheckOptions _healthCheckOptions;
+        private readonly IHealthCheckService _healthCheckService;
+
+        public HealthCheckMiddleware(RequestDelegate next, HealthCheckOptions healthCheckOptions, IHealthCheckService healthCheckService)
+        {
+            _next = next;
+            _healthCheckOptions = healthCheckOptions;
+            _healthCheckService = healthCheckService;
+        }
+
+        /// <summary>
+        /// Process an individual request.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public async Task InvokeAsync(HttpContext context)
+        {
+            // TODO: Configurable
+            if (context.Request.Path == _healthCheckOptions.Path)
+            {
+                // TODO: Caching
+                // Get results
+                var result = await _healthCheckService.CheckHealthAsync(context.RequestAborted);
+
+                // Map status to response code
+                switch (result.Status)
```

Comment:
https://httpstatuses.com/429

That is used by consul.io and others to say "backoff I am overloaded"

Reply:
I'm not sure we want to specifically use that error code, as some systems use https://httpstatuses.com/503. Health checks may not be directly related to the number of requests. Regardless: we do want to make this output somewhat configurable and even if we don't use 429 by default, we'll want to make it an option. Perhaps a dictionary mapping statuses to HTTP status codes, or some delegate handler thing.

Reply:
This being a first pass, I'll file issues for some of these issues after the PR is merged. https://github.com/aspnet/Diagnostics/issues/409

Reply:
429 is standard for say consul.io which is pretty massive. There is 200 429 and failure  (everything else and just not there). So it would be key... but sure some mapping would be fine.

Reply:
This will be a constant tension of simplicity vs configurableness. You could envision an options type that contains a map of HealthCheckStatus to status code that you could change. Assuming it didn't look terrible to use.

Reply:
My thinking would be a `Func<CompositeHealthCheckResult, int>` or something. Similarly, a `Func<HttpContext, CompositeHealthCheckResult, bool>` to determine if details should be rendered (need `HttpContext` to determine if the user is authorized, if that's how you want to do it). Then we could also provide extension methods on `HealthCheckMiddlewareOptions` to tie in with other systems (`.ShowDetailsOnlyWhenUnhealthy()` or `.ShowDetailsToAuthorizedUsersOnly(policy: "Foo")`)

Reply:
We might want to go further here and just provide a service that gets to generate the response. Since we plan on delivering integration packages for things like consul, we could ship it there and it could follow consul's conventions.

Reply:
Would we need a service or would it just be configurations/modes for the middleware? Maybe `app.UseConsulHealthChecks()`, `app.UseKubernetesHealthChecks()`, etc. Or `app.UseHealthChecks(options => options.ConfigureForConsul())`, etc.

Reply:
Ideally `UseHealthChecks()` does everything for you, and you if you want to control it you can use the options.